### DISCUSSION TITLE
I am attempting to custom-made a scroll bar. How to "Move" control?

### DISCUSSION THREAD

**Discussion by WileyNg** [picoe/Eto#2342] [General]
I have a Drawable, which has Content = someControl

I am trying to implement a scrollbar
```C#        
protected override void OnMouseWheel(MouseEventArgs e)
        {
            base.OnMouseWheel(e);
        }
```
How can I move the whole Content of the Drawable up and down using Mousewheel? without redrawing everything?

**Comment by WileyNg** [comment]
```C#
        protected override void OnLoad(EventArgs e)
        {
            _layout = new PixelLayout() { };
            _layout.Add(_control, 0, _currentHeight);
            Content = _layout;
        }

        protected override void OnMouseWheel(MouseEventArgs e)
        {
            base.OnMouseWheel(e);
            _currentHeight += (int)e.Delta.Height;
            _layout.Move(_control, new Point(0, _currentHeight));
        }


```

Not sure if this is the best method. but it feels smooth.

**Reply by allsorts46** [reply]
If it's specifically only a Drawable you want to move with the scrollbar and not other types of control, you may want to consider leaving the Drawable control where it is and just redrawing the content with a translation matrix applied instead. 

I know you say "without redrawing everything" but realistically, the drawable surface is not persistent and must be redrawn whenever the OS wants you to anyway, and you may find that at least on some platforms moving the controls' positions around might trigger layout events which cause a redraw request anyway. If double-buffering is on, the redraw should be smooth - just call Invalidate() on the control to get your OnPaint() called again as soon as possible.

If the reason for wanting to avoid drawing is that the drawing is slow, you could implement your own double buffering and only redraw the newly-exposed part when scrolling. When the scrollbar is moved, call Invalidate() on the control with the rectangle of the exposed area and cull your drawing to only parts within that area.

**Reply by WileyNg** [reply]
Thanks for the extremely lengthy response! 
Do you have any example that involves a translation matrix?
I couldn't find a good one online.

**Reply by rafntor** [reply]
This may help get started ; https://github.com/rafntor/Eto.Containers/blob/master/Eto.Containers/DragZoomImageView.cs
Another approach may be to just let a 'Scrollable' contain your 'Drawable' and get scroll-support that way?

**Reply by WileyNg** [reply]
https://user-images.githubusercontent.com/5993201/198738871-046c6f82-9bf1-4626-a1f5-78faa426cfce.mp4

I made it with PixelLayout. Feels smooth and it helps me to do other things as well such as Tooltip.

But I downloaded rafntor's example. It opens my eyes about how IMatrix can be used.
will definitely try that out next.!

**Reply by WileyNg** [reply]
https://github.com/picoe/Eto/discussions/2329#discussion-4440742
Do you guys mind reviewing some of my approach on creating a expandable UI?

I am also not sure what means by "double-buffering" and how to enable it.

**Reply by rafntor** [reply]
Looks great! Would be nice to have that Expander-class published as a nuget

**Reply by WileyNg** [reply]
> Looks great! Would be nice to have that Expander-class published as a nuget

https://github.com/picoe/Eto/discussions/2329#discussion-4440742
I posted the code there in the link above.

But like you see in the video, the animated expander is still very choppy, when there are so many controls in the panel.
Any idea how can I optimize it?
